test/e2e/gateway_api_test.go

```diff
@@ -0,0 +1,189 @@
+	featureGate := &configv1.FeatureGate{}
+	err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		name := types.NamespacedName{"default", "cluster"}
+		if err := kclient.Get(context.TODO(), name, featureGate); err != nil {
+			t.Logf("failed to get feature gate %s: %v", gwapi.FeatureGateName, err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to find feature gate: %v", err)
+	}
+
+	// Create a test namespace if needed
+	ns := buildNamespace("test-namespace")
+	if err := kclient.Create(context.TODO(), ns); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			t.Fatalf("error creating namespace: %v", err)
+		}
+	}
+	defer func() {
+		// Deleting the namespace also deletes the httproute
+		if err := kclient.Delete(context.TODO(), ns); err != nil {
+			t.Fatalf("failed to delete test namespace %s: %v", ns.Name, err)
+		}
+	}()
+
+	// Check if the feature gate is disabled, and if it is, make sure no GWAPI CRDs can be created
+	if !gwapi.FeatureIsEnabled(gwapi.FeatureGateName, featureGate) {
+		// First make sure nothing can happen with the Gateway API CRDs when the feature is disabled
+		gatewayClassFailed, gatewayFailed, httpRouteFailed := testCrds(t, ns.Name, false)
+		if !(gatewayClassFailed && gatewayFailed && httpRouteFailed) {
+			t.Fatalf("feature gate was not enabled, but gateway api objects could be created")
+		}
+		// Enable the feature gate for the rest of the test
+		featureGate.Spec.FeatureGateSelection.FeatureSet = configv1.CustomNoUpgrade
+		featureGate.Spec.FeatureGateSelection.CustomNoUpgrade = &configv1.CustomFeatureGates{
+			Enabled: []string{gwapi.FeatureGateName},
+		}
+		if err := kclient.Update(context.TODO(), featureGate); err != nil {
+			t.Fatalf("error enabling feature gate: %v", err)
+		}
+	}
+
+	// Now make sure everything can happen with the Gateway API CRDs since the feature is enabled
+	gatewayClassFailed, gatewayFailed, httpRouteFailed := testCrds(t, ns.Name, true)
+	defer func() {
+		// Delete the gateway
+		if err := kclient.Delete(context.TODO(), &gateway); err != nil {
+			t.Fatalf("failed to delete test gateway %s: %v", gateway.Name, err)
+		}
+		// Delete the gateway class
+		if err := kclient.Delete(context.TODO(), &gatewayClass); err != nil {
+			t.Fatalf("failed to delete test gatewayclass %s: %v", gatewayClass.Name, err)
+		}
+	}()
+
+	if gatewayClassFailed || gatewayFailed || httpRouteFailed {
+		t.Fatalf("feature gate was enabled, but gateway api objects could not be created")
+	}
+
+	t.Logf("successful test of gateway api crds")
+}
+
+func testCrds(t *testing.T, ns string, log bool) (gatewayClassFailed, gatewayFailed, httpRouteFailed bool) {
+	t.Helper()
+	gatewayFailed, gatewayClassFailed, httpRouteFailed = false, false, false
+
+	// Check if GatewayClass can be created.
+	gatewayClass = buildGatewayClass("openshift-default", "openshift.io/gateway-controller")
+	if err := kclient.Create(context.TODO(), &gatewayClass); err != nil {
```

Comment:
If the CR already exists, maybe consider it a success?  (Same applies for the Gateway and HTTPRoute below.)
```suggestion
	if err := kclient.Create(context.TODO(), &gatewayClass); err != nil && !errors.IsNotFound(err) {
```

Reply:
Looks like I added this then removed it.  Will look into it.

Reply:
I added this back in for the GatewayClass, of which there should be only one and it is ok if it's already created.  But the Gateway and HTTPRoute should be deleted when the test exits.